eth-contracts/contracts/CKBChain.sol

```diff
@@ -3,444 +3,163 @@ pragma solidity ^0.8.0;
+                "\x19\x01", // solium-disable-line
+                DOMAIN_SEPARATOR,
+                keccak256(abi.encode(ADD_HISTORY_TX_ROOT_TYPEHASH, _initBlockNumber, _latestBlockNumber, _historyTxRoot))
+            )
         );
-        require(
-            canonicalTransactionsRoots[blockHash] != bytes32(0),
-            "blockHash invalid or too old"
-        );
-        uint16 index = proofView.txMerkleIndex();
-        uint16 sibling;
-        uint256 lemmasIndex = 0;
-        bytes29 lemmas = proofView.lemmas();
-        uint256 length = lemmas.len() / 32;
-
-        // calc the rawTransactionsRoot
-        bytes32 rawTxRoot = proofView.txHash();
-        while (lemmasIndex < length && index > 0) {
-            sibling = ((index + 1) ^ 1) - 1;
-            if (index < sibling) {
-                rawTxRoot = Blake2b.digest64Merge(rawTxRoot, lemmas.indexH256Array(lemmasIndex));
-            } else {
-                rawTxRoot = Blake2b.digest64Merge(lemmas.indexH256Array(lemmasIndex), rawTxRoot);
-            }
 
-            lemmasIndex++;
-            // index = parent(index)
-            index = (index - 1) >> 1;
-        }
-
-        // calc the transactionsRoot by [rawTransactionsRoot, witnessesRoot]
-        bytes32 transactionsRoot = Blake2b.digest64Merge(rawTxRoot, proofView.witnessesRoot());
-        require(
-            transactionsRoot == canonicalTransactionsRoots[blockHash],
-            "proof not passed"
-        );
-        return true;
-    }
+        // 2. validatorsApprove
+        validatorsApprove(msgHash, signatures, multisigThreshold_);
 
-    // Remove hashes from the Canonical chain that are at least as old as the given header number.
-    function _gcCanonicalChain(uint64 blockNumber) internal {
-        uint64 number = blockNumber;
-        while (true) {
-            if (number == 0 || canonicalHeaderHashes[number] == bytes32(0)) {
-                break;
-            }
+        initBlockNumber = _initBlockNumber;
+        latestBlockNumber = _latestBlockNumber;
+        historyTxRoot = _historyTxRoot;
 
-            delete canonicalTransactionsRoots[canonicalHeaderHashes[number]];
-            delete canonicalHeaderHashes[number];
-            number--;
-        }
+        // 3. event
+        emit HistoryTxRootAdded(_initBlockNumber, _latestBlockNumber, _historyTxRoot);
     }
 
-    // Remove information about the headers that are at least as old as the given header number.
-    function _gcHeaders(uint64 blockNumber) internal {
-        uint64 number = blockNumber;
-        while (true) {
-            if (number == 0 || allHeaderHashes[number].length == 0) {
-                break;
-            }
-
-            for (uint256 i = 0; i < allHeaderHashes[number].length; i++) {
-                delete blockHeaders[allHeaderHashes[number][i]];
-            }
-            delete allHeaderHashes[number];
-            number--;
-        }
+    function getLatestBlockNumber() view public returns (uint64) {
```

Comment:
This is redundant because the field is public. You can get the value without this function. Check the abi file.